# <font color=blue>Assignments for "Evaluating Goodness of Fit"</font>

As in previous lessons, please submit a link to a single gist that contains links to two Juypyter notebooks (one for each assignment below).

## 1. Weather model

For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import Normalizer

%matplotlib inline
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Like in the previous lesson, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 

In [42]:
Y = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']
X = weather_df[['Humidity', 'Wind Speed (km/h)']]

In [43]:
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:55:44   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


Both R-squared and adjusted R-squared are equal and 0.288. Generally, we wan to find the R-square value is greater than or equal to 0.7. In particular, we can say that R-square values ​​below 0.3 are actually not good.

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one? 


In [44]:
X['humidity*windspeed'] = X['Humidity'] * X['Wind Speed (km/h)']
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:55:44   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity*windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


Now the R-squared values ​​are 0.341 and it is greater than the previous model.

- Add *visibility* as additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [45]:
X['visibility'] = weather_df['Visibility (km)']
X.drop('humidity*windspeed',axis=1, inplace=True) #for returning the first model we need to drop this new feature

results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:55:44   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
visibility           -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

When we compare this model to first model, both R-swuared and adjusted R-squared values increase.

- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

For both AIC and BIC, the lower value the better. Hence, the second model was better with respect to their AIC and BIC scores.

##  2. House prices model

In this exercise, you'll work on your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [46]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.

In [47]:
house = house_prices_df.drop(['Id','PoolQC','MiscFeature', 'Alley','Fence', 'FireplaceQu', 'LotFrontage'], axis=1)
house = house.drop(house[(house['GrLivArea']>4000) & (house['SalePrice']<300000)].index)
house = house.drop(house[(house["TotalBsmtSF"]>6000) & (house["SalePrice"]<200000)].index)
house = house.drop(house[(house["TotRmsAbvGrd"]>12) & (house["SalePrice"]<250000)].index)

In [48]:
house['PoolArea'] = pd.get_dummies(house['PoolArea'],drop_first=True)
house['GarageArea'] = pd.get_dummies(house['GarageArea'],drop_first=True)
house['TotalBsmtSF'] = pd.get_dummies(house['TotalBsmtSF'],drop_first=True)
house['2ndFlrSF'] = pd.get_dummies(house['2ndFlrSF'],drop_first=True)
house['Fireplaces'] = pd.get_dummies(house['Fireplaces'],drop_first=True)

In [49]:
house['GarageYrBlt'] = house['GarageYrBlt'].fillna(house['YearBuilt'])
house['MasVnrArea'] = house['MasVnrArea'].fillna(house['MasVnrArea'].mean())

In [50]:
house['SalePrice'] = np.log(house['SalePrice'])

In [51]:
from sklearn.preprocessing import LabelEncoder
categorical_feature_mask = house.dtypes==object
categorical_cols = house.columns[categorical_feature_mask].tolist()
labelencoder = LabelEncoder()
house[categorical_cols] = house[categorical_cols].apply(lambda col: labelencoder.fit_transform(col.astype(str)))

In [52]:
X = house[['OverallQual','YearBuilt','YearRemodAdd','TotRmsAbvGrd','GarageYrBlt','GarageCars','GarageArea']]

Y = house['SalePrice']

In [53]:
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.786
Model:                            OLS   Adj. R-squared:                  0.785
Method:                 Least Squares   F-statistic:                     760.2
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:55:47   Log-Likelihood:                 391.94
No. Observations:                1457   AIC:                            -767.9
Df Residuals:                    1449   BIC:                            -725.6
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            4.0280      0.624      6.457      0.000       2.804       5.252
OverallQual      0.1373      0.005     26.108      0.000       0.127       0.148
YearBuilt        0.0023      0.000      7.337      0.000       0.002       0.003
YearRemodAdd     0.0021      0.000      6.790      0.000       0.002       0.003
TotRmsAbvGrd     0.0575      0.004     16.370      0.000       0.051       0.064
GarageYrBlt     -0.0011      0.000     -2.991      0.003      -0.002      -0.000
GarageCars       0.1166      0.009     12.527      0.000       0.098       0.135
GarageArea      -0.2405      0.132     -1.827      0.068      -0.499       0.018
==============================================================================
Omnibus:                       75.937   Durbin-Watson:                   1.989
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              174.783
Skew:                          -0.306   Prob(JB):                     1.11e-38
Kurtosis:                       4.583   Cond. No.                     4.40e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.4e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

* F-statistic: 760.2
* R-squared: 0.786
* Adjusted R-squared: 0.785
* AIC: -767.9
* BIC: -725.6



- Do you think your model is satisfactory? If so, why?

R-squared and adjusted R-squared values are high enough, that is greater than  0.7 and alse F-statistic value is good enough, hence the model can be satisfactory.

- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. 

In [54]:
X = house.drop('SalePrice', axis=1)
Y = house['SalePrice']

X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.919
Model:                            OLS   Adj. R-squared:                  0.915
Method:                 Least Squares   F-statistic:                     214.4
Date:                Sun, 20 Jun 2021   Prob (F-statistic):               0.00
Time:                        23:55:47   Log-Likelihood:                 1097.9
No. Observations:                1457   AIC:                            -2048.
Df Residuals:                    1383   BIC:                            -1657.
Df Model:                          73                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const            18.0682      4.871      3.710      0.000       8.514      27.623
MSSubClass    -7.127e-05      0.000     -0.425      0.671      -0.000       0.000
MSZoning         -0.0166      0.006     -2.995      0.003      -0.028      -0.006
LotArea        2.761e-06   3.91e-07      7.063      0.000    1.99e-06    3.53e-06
Street            0.2076      0.051      4.049      0.000       0.107       0.308
LotShape         -0.0038      0.002     -1.542      0.123      -0.009       0.001
LandContour      -0.0104      0.005     -2.048      0.041      -0.020      -0.000
Utilities        -0.1300      0.122     -1.062      0.289      -0.370       0.110
LotConfig        -0.0022      0.002     -1.092      0.275      -0.006       0.002
LandSlope        -0.0070      0.014     -0.490      0.624      -0.035       0.021
Neighborhood     -0.0003      0.001     -0.604      0.546      -0.001       0.001
Condition1        0.0018      0.004      0.485      0.628      -0.006       0.009
Condition2       -0.0152      0.013     -1.206      0.228      -0.040       0.009
BldgType         -0.0113      0.005     -2.074      0.038      -0.022      -0.001
HouseStyle        0.0012      0.002      0.511      0.609      -0.004       0.006
OverallQual       0.0606      0.004     13.823      0.000       0.052       0.069
OverallCond       0.0426      0.004     11.048      0.000       0.035       0.050
YearBuilt         0.0018      0.000      5.948      0.000       0.001       0.002
YearRemodAdd      0.0006      0.000      2.446      0.015       0.000       0.001
RoofStyle         0.0006      0.004      0.153      0.879      -0.008       0.009
RoofMatl         -0.0039      0.006     -0.710      0.478      -0.015       0.007
Exterior1st      -0.0043      0.002     -2.224      0.026      -0.008      -0.001
Exterior2nd       0.0041      0.002      2.386      0.017       0.001       0.008
MasVnrType        0.0167      0.006      2.976      0.003       0.006       0.028
MasVnrArea     3.052e-05   2.24e-05      1.363      0.173   -1.34e-05    7.45e-05
ExterQual        -0.0123      0.007     -1.699      0.089      -0.027       0.002
ExterCond         0.0112      0.005      2.400      0.017       0.002       0.020
Foundation        0.0137      0.006      2.212      0.027       0.002       0.026
BsmtQual         -0.0113      0.005     -2.170      0.030      -0.022      -0.001
BsmtCond          0.0086      0.005      1.712      0.087      -0.001       0.018
BsmtExposure     -0.0050      0.003     -1.516      0.130      -0.011       0.001
BsmtFinType1     -0.0014      0.002     -0.585      0.559      -0.006       0.003
BsmtFinSF1        0.0002   1.91e-05      8.364      0.000       0.000       0.000
BsmtFinType2      0.0048      0.005      0.950      0.342      -0.005       0.015
BsmtFinSF2        0.0001   3.36e-05      4.259   

- For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

BEFORE                          
* R-squared: 0.786
* Adjusted R-squared: 0.785
* F-statistic: 760.2
* AIC: -767.9
* BIC: -725.6

AFTER
-	R-squared:	0.919
- Adj. R-squared:	0.915
- F-statistic:	214.4
- AIC:	-2048
- BIC:	-1657
	


When we compare two models, second model has the best result for us with higher r squared and adjusted r squared with smaller AIC and BIC values.